libc/unistd_file.c

```diff
@@ -0,0 +1,774 @@
+	if (fd > 0)
+		TEST_ASSERT_EQUAL_INT(0, close(fd));
+	TEST_ASSERT_EQUAL_INT(0, remove(filename));
+}
+
+
+TEST(unistd_file, file_close)
+{
+	/* close an opened descriptor */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	/* try to close a closed descriptor */
+	TEST_ASSERT_EQUAL_INT(-1, close(fd));
+	TEST_ASSERT_EQUAL_INT(EBADF, errno);
+
+	/* this test closes fd by itself, invalidate it */
+	fd = -1;
+}
+
+/* Simple usage of read() and write() */
+TEST(unistd_file, file_readwrite_nbytes)
+{
+	/* write LINE1 to file */
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+
+	/* repoen file readonly */
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+	fd = open(filename, O_RDONLY);
+	TEST_ASSERT_NOT_EQUAL_INT(-1, fd);
+	{
+		/* try to read LINE1 from file */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(fd, buf, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+
+		/* assert EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+		TEST_ASSERT_EMPTY(buf);
+	}
+}
+
+
+TEST(unistd_file, file_write_zero)
+{
+	/* test writing no bytes */
+	TEST_ASSERT_EQUAL_INT(0, write(fd, NULL, 0));
+}
+
+/* simultaneous write with two open descriptors */
+TEST(unistd_file, file_write_reopened)
+{
+	int sum;
+
+	/* reopen file descriptor and open another for reading */
+	fd2 = open(filename, O_WRONLY | O_CREAT | O_APPEND);
+	int fdr = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from two different file descriptors */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd2, LINE2, sizeof(LINE2) - 1));
+
+		/* try to read too much for simple error detection */
+		sum = (sizeof(LINE1) - 1) + (sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, read(fdr, buf, sum));
+		TEST_ASSERT_EQUAL_STRING(LINE2, buf);
+
+		/* test for EOF */
+		TEST_ASSERT_EQUAL_INT(0, read(fdr, buf, sum));
+		TEST_ASSERT_EMPTY(buf);
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fdr));
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+
+/* test simultaneous write() on duplicated file descriptor */
+TEST(unistd_file, file_write_dup)
+{
+	int sum;
+
+	/* duplicate file descriptor and open another for reading */
+	fd2 = dup(fd);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd2);
+	int fdr = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from descriptor and its copy */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd2, LINE2, sizeof(LINE2) - 1));
+
+		/* try to read too much for simple error detection */
+		sum = (sizeof(LINE1) - 1) + (sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_INT(sum, read(fdr, buf, sum));
+		TEST_ASSERT_EQUAL_STRING_LEN(LINE1, &buf[0], sizeof(LINE1) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE2, &buf[sizeof(LINE1) - 1]);
+	}
```

Comment:
No extra fd closing - fdr (certain fd leak) and fd2

Reply:
My bad. Corrected, thanks!